player/lua/select.lua

```diff
@@ -353,6 +353,101 @@ mp.add_key_binding(nil, "select-audio-device", function ()
     })
 end)
 
+local history_file_path =
+    mp.command_native({"expand-path", mp.get_property("watch-history-path")})
+
+mp.register_event("file-loaded", function ()
+    if not mp.get_property_native("save-watch-history") then
+        return
+    end
+
+    local history_file, error_message = io.open(history_file_path, "a")
+    if not history_file then
+        show_error("Failed to write the watch history: " .. error_message)
+        return
+    end
+
+    local path = mp.command_native({"normalize-path", mp.get_property("path")})
+    local title = mp.get_property("playlist/" .. mp.get_property("playlist-pos") .. "/title")
+
+    history_file:write(utils.format_json({os.time(), path, title}) .. "\n")
+    history_file:close()
+end)
+
+local function add_history_entry(line, items, paths, osd_playlist_entry)
+    local entry = utils.parse_json(line)
+
+    if not entry then
+        mp.msg.warn(line .. " in " .. history_file_path .. " is not valid JSON.")
+        return
+    end
+
+    local time, path, title = unpack(entry)
+
+    local status, date = pcall(os.date, "(%Y-%m-%d %H:%M) ", time)
+
+    if not status or not path then
+        mp.msg.warn(line .. " in " .. history_file_path .. " has invalid data.")
+        return
+    end
+
+    for i, seen_path in ipairs(paths) do
+        if seen_path == path then
+            table.remove(items, i)
+            table.remove(paths, i)
+            break
+        end
+    end
```

Comment:
I think this n^2 search will be slow. How about we keep a map for such that `map[path] = i` and get `i` if exist, and instead of removing, just update the `items[i]` with the new date?

Reply:
I did that at first but it doesn't work because i is updated as you prepend items. Also I tried doing this with the console history and it was actually much slower than looping with huge histories, maybe huge tables with long string keys cause issues.

Reply:
I tested again with 100k+ lines and performance is almost the same between looping and keeping a map. Also uploading `items[i]` as you said doesn't work because the new watched entry should be at the beginning. So while with the loop so we can just prepend items, with the map we have to decrement every `i` in the map after the removed one and revert tables at the end, that's probably slower.

Reply:
Regardless, I would still like it to be gone from reading loop.

How about we add after `for line in history_file:lines()` 
``` lua
    local seen = {}
    for i, path in ipairs(paths) do
        if seen[path] then
            table.remove(items, i)
            table.remove(paths, i)
        end
        seen[path] = true
    end
```

Reply:
> You should really use a BST to store history
I was thinking about it, but it still needs flattened to table for input.select. And also the input is already sorted and we only need to reverse it, so having simple ordered table for storage is appropriate. We only need to filter it, by path, not by date. So any sort/search capabilities by date are not in-fact needed.

I'm thinking we should, read the file in reverse order and just skip seen entries. Nothing more is needed.

```
    local lines = {}
    for line in file:lines() do
        table.insert(lines, line)
    end
    file:close()

    for i = #lines, 1, -1 do
        add_history_entry(lines[i], items, paths, osd_playlist_entry)
    end
```

and in `add_history_entry` just do `if seen[path] continue`.

Reply:
> You should really use a BST to store history

Or you meant to store it to file? In which case we may use sqlite already for better storage. We could migrate watch-later there too and future console history.

Reply:
Nevermind I was still prepending to the tables. It's 3 faster even with unique lines by just replacing prepending with appending.

Reply:
Currently it is reloaded on each search.

Do you think added complexity is worth it? After first file load it would be cached by the kernel anyway, so consecutive loads should be fast, except maybe next time after it is evicted. Keeping local state of history is valid, but not sure it is worth the added complexity.